File: vscode-extension/src/protocol.ts
/**
 * VSSync IPC Protocol Types
 * Shared type definitions for communication between VS Code and Visual Studio
 */

export type IdeType = 'vscode' | 'visualstudio';

export type MessageType = 
    | 'DISCOVER'
    | 'DISCOVER_RESPONSE'
    | 'OPEN_FILE'
    | 'OPEN_FILE_RESPONSE'
    | 'PING'
    | 'PONG';

export interface IpcMessage {
    type: MessageType;
    payload: Record<string, unknown>;
    timestamp: number;
    sourceIde: IdeType;
    sourcePid: number;
}

export interface DiscoverPayload {
    workspacePath: string;
}

export interface DiscoverResponsePayload {
    port: number;
    ide: IdeType;
    version: string;
    workspacePath: string;
    solutionPath?: string;
    pid: number;
    windowHandle?: number;
}

export interface OpenFilePayload {
    filePath: string;
    line?: number;
    column?: number;
    focus: boolean;
}

export interface OpenFileResponsePayload {
    success: boolean;
    error?: string;
}

export interface IdeInstance {
    port: number;
    ide: IdeType;
    version: string;
    workspacePath: string;
    solutionPath?: string;
    pid: number;
    windowHandle?: number;
}

export function createMessage(
    type: MessageType,
    payload: Record<string, unknown>,
    sourceIde: IdeType
): IpcMessage {
    return {
        type,
        payload,
        timestamp: Date.now(),
        sourceIde,
        sourcePid: process.pid
    };
}

export function parseMessage(data: string): IpcMessage | null {
    try {
        // Remove BOM (Byte Order Mark) and trim whitespace/control characters
        const cleanedData = data.replace(/^\uFEFF/, '').trim();
        const parsed = JSON.parse(cleanedData);
        if (parsed && typeof parsed.type === 'string' && parsed.payload) {
            return parsed as IpcMessage;
        }
        return null;
    } catch {
        return null;
    }
}

/**
 * Type guard for DiscoverPayload
 */
export function isDiscoverPayload(payload: unknown): payload is DiscoverPayload {
    if (typeof payload !== 'object' || payload === null) {
        return false;
    }
    const p = payload as Record<string, unknown>;
    return typeof p.workspacePath === 'string';
}

/**
 * Type guard for DiscoverResponsePayload
 */
export function isDiscoverResponsePayload(payload: unknown): payload is DiscoverResponsePayload {
    if (typeof payload !== 'object' || payload === null) {
        return false;
    }
    const p = payload as Record<string, unknown>;
    return typeof p.port === 'number' &&
           (p.ide === 'vscode' || p.ide === 'visualstudio') &&
           typeof p.version === 'string' &&
           typeof p.workspacePath === 'string' &&
           typeof p.pid === 'number';
}

/**
 * Type guard for OpenFilePayload
 */
export function isOpenFilePayload(payload: unknown): payload is OpenFilePayload {
    if (typeof payload !== 'object' || payload === null) {
        return false;
    }
    const p = payload as Record<string, unknown>;
    return typeof p.filePath === 'string' &&
           (p.line === undefined || typeof p.line === 'number') &&
           (p.column === undefined || typeof p.column === 'number') &&
           typeof p.focus === 'boolean';
}

/**
 * Type guard for OpenFileResponsePayload
 */
export function isOpenFileResponsePayload(payload: unknown): payload is OpenFileResponsePayload {
    if (typeof payload !== 'object' || payload === null) {
        return false;
    }
    const p = payload as Record<string, unknown>;
    return typeof p.success === 'boolean';
}

File: visual-studio-extension/VSSync/Protocol.cs
using System;
using Newtonsoft.Json;

namespace VSSync
{
    /// <summary>
    /// IPC Message Types
    /// </summary>
    public enum MessageType
    {
        DISCOVER,
        DISCOVER_RESPONSE,
        OPEN_FILE,
        OPEN_FILE_RESPONSE,
        PING,
        PONG
    }

    /// <summary>
    /// IDE Type
    /// </summary>
    public enum IdeType
    {
        vscode,
        visualstudio
    }

    /// <summary>
    /// Base IPC Message structure
    /// </summary>
    public class IpcMessage<T>
    {
        [JsonProperty("type")]
        public string Type { get; set; } = string.Empty;

        [JsonProperty("payload")]
        public T Payload { get; set; }

        [JsonProperty("timestamp")]
        public long Timestamp { get; set; }

        [JsonProperty("sourceIde")]
        public string SourceIde { get; set; } = string.Empty;

        [JsonProperty("sourcePid")]
        public int SourcePid { get; set; }

        public static IpcMessage<Tmessage> Create<Tmessage>(MessageType type, Tmessage payload, IdeType ide)
        {
            return new IpcMessage<Tmessage>
            {
                Type = type.ToString(),
                Payload = payload,
                Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                SourceIde = ide.ToString(),
                SourcePid = System.Diagnostics.Process.GetCurrentProcess().Id
            };
        }
    }

    /// <summary>
    /// Discover request payload
    /// </summary>
    public class DiscoverPayload
    {
        [JsonProperty("workspacePath")]
        public string WorkspacePath { get; set; } = string.Empty;
    }

    /// <summary>
    /// Discover response payload
    /// </summary>
    public class DiscoverResponsePayload
    {
        [JsonProperty("port")]
        public int Port { get; set; }

        [JsonProperty("ide")]
        public string Ide { get; set; } = "visualstudio";

        [JsonProperty("version")]
        public string Version { get; set; } = string.Empty;

        [JsonProperty("workspacePath")]
        public string WorkspacePath { get; set; } = string.Empty;

        [JsonProperty("solutionPath")]
        public string? SolutionPath { get; set; }

        [JsonProperty("pid")]
        public int Pid { get; set; }

        [JsonProperty("windowHandle")]
        public long WindowHandle { get; set; }
    }

    /// <summary>
    /// Open file request payload
    /// </summary>
    public class OpenFilePayload
    {
        [JsonProperty("filePath")]
        public string FilePath { get; set; } = string.Empty;

        [JsonProperty("line")]
        public int? Line { get; set; }

        [JsonProperty("column")]
        public int? Column { get; set; }

        [JsonProperty("focus")]
        public bool Focus { get; set; } = true;
    }

    /// <summary>
    /// Open file response payload
    /// </summary>
    public class OpenFileResponsePayload
    {
        [JsonProperty("success")]
        public bool Success { get; set; }

        [JsonProperty("error")]
        public string? Error { get; set; }
    }

    /// <summary>
    /// IDE Instance information
    /// </summary>
    public class IdeInstance
    {
        public int Port { get; set; }
        public IdeType Ide { get; set; }
        public string Version { get; set; } = string.Empty;
        public string WorkspacePath { get; set; } = string.Empty;
        public string? SolutionPath { get; set; }
        public int Pid { get; set; }
        public IntPtr WindowHandle { get; set; }
    }
}
